<div style="border:solid Chocolate 2px; padding: 40px">

<b> Комментарий ревьюера </b>

Настя, вижу, что в твоём проекте нет задач проекта - выполнен только осмотр данных, но сами задания не решены. Возвращаю тебе проект без проверки - дополни его, пожалуйста, заданиями из условий проекта. 
    
Проверку продолжим, когда он будет выполнен целиком.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Очень странно, они не сохранились(
Проверила еще раз перед отправкой.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Иногда платформа не сохраняет изменения, такое бывает:)

# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Как выполнить задание?

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Хорошее вступление!
    
В нём есть всё, что необходимо, чтобы понять суть проекта с первых строк отчёта!

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Библиотеки импортировали, коннектор сделали - отлично!

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
Блок загрузки библиотек лучше размещать в отдельной от остального кода ячейке. Такой подход позволит улучшить читаемость кода, снизит время поиска ошибок (если они есть) и исключит повторное выполнение кода.
        </div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 

Я вижу, исполнение твоего кода начинается в ячейке с номером, который отличается от единицы. Перед отправкой проекта стоит проверять работоспособность кода — это можно сделать, нажав на панели Jupiter Hub ``Kernel`` и ``Restart & Run All`` (см скриншот ниже).

![](https://i.postimg.cc/yd19rYf6/Screenshot-428.png)
        
Важно также убедиться, что все ячейки проекта исполнились - можно просто пролистать работу до конца и убедиться, что последняя ячейка исполнена. Такая проверка поможет тебе убедиться, что твоё решение будет должным образом воспроизведено во время ревью или во время передачи его заказчику - это очень важный этап, пренебрегать которым не стоит:)

<div class="alert alert-info">
Поняла
</div>

**Таблица books**

In [2]:
query = '''
            SELECT * 
            FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


**Таблица authors**

In [4]:
query = '''
            SELECT * 
            FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


**Таблица ratings**

In [6]:
query = '''
            SELECT * 
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


**Таблица reviews**

In [8]:
query = '''
            SELECT * 
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Данные осмотрели - отлично!

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
Есть дублирующийся код - ты выполняешь один и тот же запрос, меняя названия таблиц. Давай попробуем организовать этот процесс через цикл? Подставлять название таблицы в строку-запрос можно с помощью `f-строки` или метода `format()`. Саму строку-запрос можно создавать внутри цикла на каждой итерации.

</div>

<div class="alert alert-info">
Поняла, спасибо
</div>

# Ответы на вопросы

## Сколько книг вышло после 1 января 2000 года?

In [9]:
query = '''
            SELECT COUNT(book_id) count_of_books
            FROM books 
            WHERE publication_date > '01-01-2000'
        '''
pd.io.sql.read_sql(query, con = engine)

count_of_books
0             819

После 1 января 2000 года было выпущено 819 книг.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно!
    
Отмечу, что условие фильтрации `publication_date >= '2000-01-01' ` также было бы правильным, хоть и привело бы к другому результату - в реальных проектах о том, включается ли дата в диапазон или нет, всегда стоит уточнять заранее:)

</div>

## Для каждой книги посчитайте количество обзоров и среднюю оценку

In [10]:

query = '''
            SELECT b.title, rat.avg_rating, rev.review_count
            FROM books b
            LEFT JOIN (SELECT book_id, AVG(rating) avg_rating FROM ratings GROUP BY book_id) rat ON b.book_id=rat.book_id
            LEFT JOIN (SELECT book_id, COUNT(*) review_count FROM reviews GROUP BY book_id) rev ON b.book_id=rev.book_id
        '''
pd.io.sql.read_sql(query, con = engine)

title  avg_rating  \
0            The Body in the Library (Miss Marple  #3)    4.500000   
1                                            Galápagos    4.500000   
2                             A Tree Grows in Brooklyn    4.250000   
3    Undaunted Courage: The Pioneering First Missio...    4.000000   
4                                          The Prophet    4.285714   
..                                                 ...         ...   
995                                Alice in Wonderland    4.230769   
996          A Woman of Substance (Emma Harte Saga #1)    5.000000   
997                                          Christine    3.428571   
998  The Magicians' Guild (Black Magician Trilogy  #1)    3.500000   
999                           The Plot Against America    3.000000   

     review_count  
0             2.0  
1             2.0  
2             5.0  
3             2.0  
4             4.0  
..            ...  
995           4.0  
996           2.0  
997           3.0  
998           2.0  
999           2.0  

[1000 rows x 3 columns]

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Метрики для книг посчитаны верно! Здорово, что метрики посчитаны и для тех книг, для которых нет отзывов - довольно частая ошибка в этом задании:)

</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Для идентификации книг вместо `id` стоит выводить на экран название книги.

<div class="alert alert-info">
Исправила
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено.

Для большинства книг было написано несколько обзоров и средние оценки варьируются.

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [11]:
query = '''
            SELECT b.publisher_id, p.publisher, COUNT(*) count_books
            FROM books b
            LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY b.publisher_id, p.publisher
            ORDER BY COUNT(*) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  count_books
0           212  Penguin Books           42

Издательство "Penguin Books" выпустило 42 книги толще 50 страниц. Это около 4% от всех книг в данных.

Это Британское издательство, основанное в 1935 году в Лондоне сэром Алленом Лейном и его братьями Ричардом и Джоном. Главной заслугой издательства считают демократизацию книжного рынка, превратившую книгу из предмета роскоши в удовольствие, доступное массам. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Издательство опредили верно! Спасибо, что дополнительно выведено количество книг издательства:)

</div>

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [12]:
query = '''
            SELECT a.author, AVG(ratings)
            FROM (
            SELECT book_id, AVG(rating) ratings
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(*) > 50
            ) rat
            JOIN (SELECT book_id, author_id FROM books) b ON rat.book_id=b.book_id
            JOIN (SELECT author_id, author FROM authors) a ON b.author_id=a.author_id
            GROUP BY a.author
            ORDER BY avg(ratings) DESC

            
        '''
pd.io.sql.read_sql(query, con = engine)

author       avg
0                          J.K. Rowling/Mary GrandPré  4.283844
1                   Markus Zusak/Cao Xuân Việt Khương  4.264151
2                                      J.R.R. Tolkien  4.258446
3                                   Louisa May Alcott  4.192308
4                                        Rick Riordan  4.080645
5                                     William Golding  3.901408
6                                       J.D. Salinger  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
9                                           Dan Brown  3.754540
10                                         Lois Lowry  3.750000
11            George Orwell/Boris Grabnar/Peter Škerl  3.729730
12                                    Stephenie Meyer  3.662500
13                                     John Steinbeck  3.622951

Автор с самой высокой оценкой - Джоан Роулинг, автор книг о Гарри Поттере. Тут она представлена в соавторстве с иллюстратором Мэри Гранпрэ. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Автор с самым высоким средним рейтингом найден верно! Классно, что дополнительно выведен сам средний рейтинг - именно в его расчёте обычно делают ошибки:)

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
В задании сказано, что нужно найти одного автора, поэтому было бы логичнее оставить в выводе только одну строку:)

<div class="alert alert-info">
Поняла, спасибо
</div>

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [13]:
query = '''
            SELECT AVG(count_reviews)
            FROM (
            SELECT username, COUNT(*) count_reviews
            FROM reviews
            WHERE username IN
            (
            SELECT username--, COUNT(*)
            FROM ratings
            GROUP BY username
            HAVING COUNT(*) > 48
            )
            GROUP BY username) rev
        

        '''
pd.io.sql.read_sql(query, con = engine)


avg
0  24.0

Среднее количество обзоров от пользователей, которые поставили более 48 оценок - 24.
Видимо люди, которые любят читать, так же любят поделиться своим мнением с другими любителями чтения)

<div class="alert alert-info">
Исправила. В пачке было 50 почему-то
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Это задание списано: в актуальной версии задания в условии сказано про более 48 отзывов, ты используешь условие более 50 отзывов из старой версии условий. Нужно переделать с актуальным условием.
    
Дополнительные задания для тебя:
1. Выведи таблицу, которая будет содержать по году публикации (отобразить нужно только те года, в которых издано более 30 книг):

- количество издательств,
- выпущенных книг и
- сколько всего тысяч страниц было в изданных книгах.
2. Выведи в одной таблице два числа — среднюю оценку тех книг, на которые написало отзывов более 3 человек и отдельно среднюю оценку остальных книг, сделай выводы, какой рейтинг больше.

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено.

Выведи таблицу, которая будет содержать по году публикации (отобразить нужно только те года, в которых издано более 30 книг):
- количество издательств,
- выпущенных книг и
- сколько всего тысяч страниц было в изданных книгах.

In [14]:
query = '''
            SELECT EXTRACT(YEAR FROM publication_date) years,
            COUNT (DISTINCT publisher_id) publishers, 
            COUNT (book_id) books,
            to_char (SUM (num_pages), '99G999D')  pages
            FROM books 
            GROUP BY 1
            HAVING COUNT (*) > 30
        '''
pd.io.sql.read_sql(query, con = engine)

years  publishers  books    pages
0  1999.0          26     41   15,763
1  2000.0          35     38   13,328
2  2001.0          41     60   21,758
3  2002.0          62     94   38,597
4  2003.0          65    105   41,423
5  2004.0          88    124   46,779
6  2005.0          89    139   55,967
7  2006.0         109    184   68,302
8  2007.0          38     50   18,258

Выведи в одной таблице два числа — среднюю оценку тех книг, на которые написало отзывов более 3 человек и отдельно среднюю оценку остальных книг, сделай выводы, какой рейтинг больше.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
В целом верно, но страницы должны быть в тысячах.

<div class="alert alert-info">
Теперь в тысячах
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.3 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено.

<div class="alert alert-info">
Тут я пыталась всю дорогу, не знаю как вывести два числа в одной таблице. По отдельности получилось.
У меня еще дедлайн вчера был, не знаю что делать. Куратор не отвечает(
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Сегодня ещё можно сдавать. 
   
Сам запрос в целом неправильный:
1. Операторы сравнения не соответствуют условию.
2. Нет группировки, чтобы работал `HAVING`.
3. Фильтровать нужно не рейтинг, а книги, у которых нужное количество отзывово - по этим книгам нужно сделать фильтрацию при оценке среднего рейтинга.

Выведи в одной таблице два числа — среднюю оценку тех книг, на которые написало отзывов более 3 человек и отдельно среднюю оценку остальных книг, сделай выводы, какой рейтинг больше. 
- mean rating
- where COUNT rating_id > 3 
- mean rating  rest books

<div class="alert alert-info">
СДЕЛАНО. У меня нет интернета, я не в России. Поэтому захожу в Мак и Старбакс каждый раз для правок.
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.3 <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Запрос теперь ближе к правде, но при оценке рейтинга ты в итоговую таблицу сразу добавляешь и рейтинги, и отзывы, в результате у тебя появляются связи `многие ко многим`, и значение рейтинга немного смещается. 
    
Более правильным подходом было бы сначала найти книги, которые соответсвуют условию фильтрации про количество отзывов, а потом уже считать рейтинг, выполняя фильтрацию по нужным книгам, например, через `WHERE` или `INNER JOIN`.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.4 <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Ещё на шаг ближе к правде. Обрати, пожалуйста, внимание, что ты всё ещё в одну и ту же таблицу соединяешь и рейтинги, и отзывыв - в результате ты получаешь дублирование столбцов. 
    
Как я писал ранее, тут можно:
- сначала отдельным запросом найти книги, которым поставили нужное количество оценок;
- далее при расчёте рейтнга фильтровать эти книги - проще через `WHERE`, `INNER JOIN` можно использовать для фильтрации книг, но не для расчёта самого рейтинга из-за дубликатов.
    
В итоге скелет твоего запроса может выглядеть так:
    
    WITH true_books AS (
        
        # здесь запрос, который выводит ID книг, у которых более 3 отзывов
  
        ),
        
    more_3 AS ( --тут запрос, который фильтрует книги и считает рейтинг
       
        ...
        WHERE book_id IN ()
        ),
    less_eq_3 AS ( --тут запрос, который фильтрует книги и считает рейтинг
    
    WHERE book_id NOT IN ()
    )
    
    SELECT * FROM ... --итоговый запрос
    

<div class="alert alert-info">
у меня украли зарядку от ноута, в Москве я буду только через 2 недели. Я очень устала, если честно, править эту доп задачу в 4 раз. и возможности продолжать это у меня нет.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.5 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Текущий запрос не исполняется, так как содержит ошибки. На этой итерации можем закончить - я разберу решение задачи, чтобы было понятно, как её решать. Ниже - мой запрос с комментариями.

In [15]:
query = '''

WITH many_rew AS ( --в этом блоке запроса many_rew ищем книги, у которых более 3 отзыва
    SELECT 
        book_id,
        COUNT(DISTINCT username) as rews
    FROM reviews
    GROUP BY book_id
    HAVING COUNT(DISTINCT username) > 3
)
SELECT AVG(rating) as rating -- в этом блоке запроса считаем рейтинг отдельно по книгам, у которых более 3 отзывов
FROM ratings
WHERE ratings.book_id IN (SELECT book_id FROM many_rew)

UNION ALL -- с помощью UNION ALL соединяем два числа

SELECT AVG(rating) as rating -- в этом блоке счтитаем рейтинг отдельно по книгам, которые не попали в предыдущий расчёт
FROM ratings
WHERE ratings.book_id NOT IN (SELECT book_id FROM many_rew)
'''

pd.io.sql.read_sql(query, con = engine)

rating
0  3.949505
1  3.890375

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 
По завершении проекта стоит сделать итоговый вывод: в него можно включить основные шаги проекта, а также ключевые выводы, полученные в ходе работы над ним.
        </div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено.

# Общий вывод


1. Описали цели исследования;
2. Исследовали таблицы; 
3. Сделали по одному SQL-запросу для решения каждого задания;
- После 1 января 2000 года было выпущено 819 книг
- Для каждой книги посчитали количество обзоров и среднюю оценку
- Определили издательство, которое выпустило наибольшее число книг толще 50 страниц - Издательство "Penguin Books" выпустило 42 книги толще 50 страниц. Это около 4% от всех книг в данных.
- Определили автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками - Джоан Роулинг, автор книг о Гарри Поттере. Тут она представлена в соавторстве с иллюстратором Мэри Гранпрэ. 
-  Посчитайли среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.
4. Вывели результат каждого запроса в тетрадке;
5. Описали выводы по каждой из решённых задач.


<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту.

Настя, продолжаем работу над проектом.

Есть небольшие рекомендации по улучшению проекта - некритичная мелочь, которая не займёт много времени, но сделает твой проект стилистически более совершенным. Будет классно, если эти рекомендации будут учтены в твоей работе:)
    
К нескольким заданиям стоит вернуться - к ним я оставил комментарии красного цвета с пояснениями внутри. Думаю, ты быстро разберёшься, в чём дело.
    
Также нужно решить дополнительные задания.

Буду рад ответить на любые твои вопросы по проекту или на какие-либо другие, если они у тебя имеются - оставь их в комментариях, и я постараюсь ответить:)
    
Жду твой проект на повторном ревью. До встречи:)

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту v.2.

Настя, продолжаем работу над проектом - актуальные замечания отмечены комментариями с меткой `v.2`.
    
Жду тебя снова:)

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту v.3.

Настя, продолжаем работу над проектом - актуальные замечания отмечены комментариями с меткой `v.3`.
    
Жду тебя снова:)

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту v.4.

Настя, продолжаем работу над проектом - актуальные замечания отмечены комментариями с меткой `v.4`.
    
Жду тебя снова:)

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту v.5.

Настя, все замечания учтены - проект принят!
    
Спасибо за работу над проектом, желаю приятного завершения выпускного модуля:)